sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -160,34 +167,46 @@ class CacheManager extends Logging {
   /**
    * Tries to re-cache all the cache entries that refer to the given plan.
    */
-  def recacheByPlan(spark: SparkSession, plan: LogicalPlan): Unit = writeLock {
+  def recacheByPlan(spark: SparkSession, plan: LogicalPlan): Unit = {
     recacheByCondition(spark, _.find(_.sameResult(plan)).isDefined)
   }
 
   private def recacheByCondition(
       spark: SparkSession,
       condition: LogicalPlan => Boolean,
       clearCache: Boolean = true): Unit = {
-    val it = cachedData.iterator()
     val needToRecache = scala.collection.mutable.ArrayBuffer.empty[CachedData]
-    while (it.hasNext) {
-      val cd = it.next()
-      if (condition(cd.plan)) {
-        if (clearCache) {
-          cd.cachedRepresentation.cacheBuilder.clearCache()
+    writeLock {
+      val it = cachedData.iterator()
+      while (it.hasNext) {
+        val cd = it.next()
+        if (condition(cd.plan)) {
+          // Remove the cache entry before we create a new one, so that we can have a different
+          // physical plan.
+          needToRecache += cd
+          it.remove()
+        }
+      }
+    }
+    val recomputedPlans = needToRecache.map { cd =>
+      if (clearCache) {
+        cd.cachedRepresentation.cacheBuilder.clearCache()
+      }
+      val plan = spark.sessionState.executePlan(cd.plan).executedPlan
+      val newCache = InMemoryRelation(
+        cacheBuilder = cd.cachedRepresentation.cacheBuilder.withCachedPlan(plan),
+        logicalPlan = cd.plan)
+      cd.copy(cachedRepresentation = newCache)
+    }
+    writeLock {
```

Comment:
instead of locking the entire loop, we can make the lock per-loop
```
needToRecache.foreach { cd =>
  if (clearCache) {
    cd.cachedRepresentation.cacheBuilder.clearCache()
  }
  val plan = spark.sessionState.executePlan(cd.plan).executedPlan
  val newCache = InMemoryRelation(
    cacheBuilder = cd.cachedRepresentation.cacheBuilder.withCachedPlan(plan),
    logicalPlan = cd.plan)
  writeLock {
    cachedData.add(cd.copy(cachedRepresentation = newCache))
  }
}
```